Allow nn.ChannelShuffle to run without erroring on CUDA tensors

isdanni opened the pull request:
Summary: Include GPU support for `nn.ChannelShuffle` & update test.

Fix: #104603

Test Plan: Please see GitHub Actions.

Differential Revision: D47523764

isdanni commented:
Could you please review this diff? :) 

cc @kimishpatel  @mikaylagawarecki

mikaylagawarecki commented on test/test_nn.py:
```diff
@@ -6490,6 +6490,10 @@ def test_affine_grid_3d(self):
                 self.assertEqual(input_cpu.grad, input_gpu.grad)
 
     def test_channel_shuffle(self):
+        device_list = ['cpu']
```
you can move this test to [`TestNNDeviceType`](https://github.com/pytorch/pytorch/blob/7823b3501538ddfeab0ebeb79bdd444212715202/test/test_nn.py#L7928) to parameterize over device rather than using a `device_list` here

mikaylagawarecki reviewed (commented):
Thanks @isdanni for the fix. I modified the PR title slightly to avoid giving the impression that a CUDA kernel had been added for ChannelShuffle :)

I'm not sure whether you have any interest here but I am curious whether you would be interested in adding a ModuleInfo for `torch.nn.ChannelShuffle`.

At a high level, [`ModuleInfo`s](https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_modules.py#L2243) allow a more centralized testing framework for `nn.Module`, [each ModuleInfo in the module_db over here](https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_modules.py#L2243) has an associated sample_inputs function and tests in [test_modules.py](https://github.com/pytorch/pytorch/blob/main/test/test_modules.py) are run over each the sample_inputs for each module. 

I can provide more detail if you would be interested. But for this PR just moving the test to the `TestNNDeviceType` class is sufficient :)

isdanni committed:
Allow nn.ChannelShuffle to run without erroring on CUDA tensors (#105351)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/105351

Include GPU support for `nn.ChannelShuffle` & update test.

Test Plan: Please see GitHub Actions.

Differential Revision: D47523764

fbshipit-source-id: 70c6c7e385386adf3b2dcec25003ab42a3ed0121

isdanni commented:
> Thanks @isdanni for the fix. I modified the PR title slightly to avoid giving the impression that a CUDA kernel had been added for ChannelShuffle :)
> 
> I'm not sure whether you have any interest here but I am curious whether you would be interested in adding a ModuleInfo for `torch.nn.ChannelShuffle`.
> 
> At a high level, [`ModuleInfo`s](https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_modules.py#L2243) allow a more centralized testing framework for `nn.Module`, [each ModuleInfo in the module_db over here](https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_modules.py#L2243) has an associated sample_inputs function and tests in [test_modules.py](https://github.com/pytorch/pytorch/blob/main/test/test_modules.py) are run over each the sample_inputs for each module.
> 
> I can provide more detail if you would be interested. But for this PR just moving the test to the `TestNNDeviceType` class is sufficient :)

Thanks for the quick review! I moved it under `TestNNDeviceType` and added `@skipMeta`(Since we are only testing CPU and CUDA atm).

I'm interested in adding ModuleInfo for `nn.ChannelShuffle`. Will send a separate PR soon. And yes a bit extra information would definitely be helpful :)

mikaylagawarecki reviewed (approved):
Thanks!

kimishpatel commented:
Just for my understanding, there is no cuda impl for channel shuffle in this diff. Then how is it allowing channel shuffle to run when input tensors are cuda tensors? Are you converting them to cpu tensors somewhere?

mikaylagawarecki commented:
@kimishpatel adding 
```
dispatch:
    CPU, CUDA: cpu_impl
```
in native_functions allows the CPU impl to be run for CUDA tensors is my understanding, so this doesn't actually add a CUDA impl but just prevents ChannelShuffle from erroring out if it is passed inputs that are on CUDA. Does that sound reasonable to you?

kimishpatel commented:
> @kimishpatel adding
> 
> ```
> dispatch:
>     CPU, CUDA: cpu_impl
> ```
> 
> in native_functions allows the CPU impl to be run for CUDA tensors is my understanding, so this doesn't actually add a CUDA impl but just prevents ChannelShuffle from erroring out if it is passed inputs that are on CUDA. Does that sound reasonable to you?

Right. I understood that, but what I dont understand is what happens when you actually dispatch to CPU impl. Shouldnt it error out then?
